### PR TITLE
feature: record span durations as a metric

### PR DESCRIPTION
**Pull request by emberian**
this isn't quite #770, but it is probably a lot of metric.

### CODE REVIEW

File: rust/chains/optics-ethereum/src/settings.rs [hunk 1]
```diff
@@ -86,6 +86,14 @@ pub enum EthereumConnection {
     },
 }
 
+impl Default for EthereumConnection {
+    fn default() -> Self {
+        EthereumConnection::Ws {
+            url: "wss://main-light.eth.linkpool.io/ws".into(),
```
Comment (yourbuddyconner) [inline rust/chains/optics-ethereum/src/settings.rs:92]: What does this do?

**Comment by yourbuddyconner** [comment]
```
# HELP optics_span_duration_sec Duration from span creation to span destruction
# TYPE optics_span_duration_sec histogram
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="0.005"} 5
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="0.01"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="0.025"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="0.05"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="0.1"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="0.25"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="0.5"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="1"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="2.5"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="5"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="10"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_base::home",le="+Inf"} 6
optics_span_duration_sec_sum{VERSION="0.1.0",span_name="produce_update",target="optics_base::home"} 0.02409075
optics_span_duration_sec_count{VERSION="0.1.0",span_name="produce_update",target="optics_base::home"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="0.005"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="0.01"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="0.025"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="0.05"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="0.1"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="0.25"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="0.5"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="1"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="2.5"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="5"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="10"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home",le="+Inf"} 6
optics_span_duration_sec_sum{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home"} 0.023828915
optics_span_duration_sec_count{VERSION="0.1.0",span_name="produce_update",target="optics_ethereum::home"} 6
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="0.005"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="0.01"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="0.025"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="0.05"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="0.1"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="0.25"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="0.5"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="1"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="2.5"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="5"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="10"} 1
optics_span_duration_sec_bucket{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home",le="+Inf"} 1
optics_span_duration_sec_sum{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home"} 0.002599579
optics_span_duration_sec_count{VERSION="0.1.0",span_name="updater",target="optics_ethereum::home"} 1
```

  - I notice on the updater target is different but the span `produce_update` is duplicated. Not sure if intentional. 
- Could we add an `agent` label to these metrics?

File: rust/chains/optics-ethereum/src/settings.rs (hunk 1 above)
Comment (emberian) [reply rust/chains/optics-ethereum/src/settings.rs:92]: No idea, seems like a sane default though.

Comment (emberian) [reply rust/chains/optics-ethereum/src/settings.rs:92]: What's the easy way to start using `optics-ethereum` without this impl?

Comment (yourbuddyconner) [reply rust/chains/optics-ethereum/src/settings.rs:92]: Perhaps putting this in the docs as a nice default, but I think we should leave it up to the user to get their own RPC endpoint.

Comment (emberian) [reply rust/chains/optics-ethereum/src/settings.rs:92]: my opinion: docs aren't defaults, docs are just writing. `Default` impls are defaults. path of least resistance becomes code... is the concern that the default will leak into prod if it's in code and not in a dead file?

Comment (prestwich) [reply rust/chains/optics-ethereum/src/settings.rs:92]: I think I'd rather have a compilation error than a default?

Comment (yourbuddyconner) [reply rust/chains/optics-ethereum/src/settings.rs:92]: >  docs aren't defaults, docs are just writing. Default impls are defaults.
I hear what you're saying here @emberian, this change does make it easier for an end-user running our software. 

>  is the concern that the default will leak into prod if it's in code and not in a dead file?
No, including a node service as a default that we as a team have no control over and have done no vetting of implies some level of endorsement by us. I am not willing to endorse this service on behalf of the Optics core team, despite it being useful. 

We can include cLabs services like Forno, as we have control over it, but _really_, setting up a node (or making the active decision to use a free node provider like linkpool.io) is a pre-requisite and should be an informed choice by the end-user.

File: rust/agents/kathy/src/main.rs [hunk 2]
```diff
@@ -16,11 +16,15 @@ use crate::{kathy::Kathy, settings::KathySettings as Settings};
...
-    settings.base.tracing.start_tracing()?;
 
     let agent = Kathy::from_settings(settings).await?;
 
+    agent
+        .as_ref()
+        .settings
+        .tracing
+        .start_tracing(agent.metrics().span_duration())
+        .expect("tracing is unhappy");
```
Comment (prestwich) [inline rust/agents/kathy/src/main.rs:27]: why does this behavior differ from other agents?

File: rust/chains/optics-ethereum/src/settings.rs (hunk 1 above)
Comment (prestwich) [reply rust/chains/optics-ethereum/src/settings.rs:92]: Gonna make a call on this: The agent should be 100% BYO Node. We should make it non-functional unless the user sepcifically configures a node RPC. Go ahead and delete this line and add a doc with suggestions and disclaimer that we haven't verified that this URL is suitable for use

File: rust/optics-base/src/settings/mod.rs [hunk 3]
```diff
@@ -155,6 +155,20 @@ pub struct Settings {
     pub signers: HashMap<String, SignerConf>,
 }
 
+impl Settings {
+    /// Private so that
```
Comment (prestwich) [inline rust/optics-base/src/settings/mod.rs:159]: I'm in suspense here :)

**Comment by prestwich** [comment]
resolve conflicts and lets get this in :)

File: rust/agents/kathy/src/main.rs (hunk 2 above)
Comment (emberian) [reply rust/agents/kathy/src/main.rs:27]: for no good reason.

**Comment by emberian** [comment]
> Could we add an agent label to these metrics?

discussed this a bit offline but the usual agent name problem pops up -- callees don't know that somewhere above them is an agent and that they should be caring about its name. we'll hit this again if/once `ethers` has its own metrics.

**Review by prestwich — APPROVED** [review]

**Pull request merged**